### DISCUSSION TITLE
Compile SDL2 with libdecor support.

### DISCUSSION THREAD

**Discussion by RushingAlien** [ppy/osu#16693] [Ideas]
Hello, currently, Osu-lazer Appimage when run as a native Wayland client in GNOME, Weston and Enlightenment Wayland session is missing window decorations (titlebar) in windowed mode, this is due to the fact that these sessions don't draw titlebars for an app, but rather expects an app to draw its own titlebar/window decoration. This is exactly what libdecor is for, for Osu-lazer to draw its own window decorations (CSD/Client Side Decorations), and don't worry it prioritizes system drawn window decorations if available (SSD/Server Side Decorations).

And considering the next SDL2 stable will make Wayland the default backend and display server, I think this is important

**Comment by bdach** [comment]
hopefully this is something that will automatically happen as our SDL fork is updated with upstream and not require any explicit action on our behalf.

**Reply by RushingAlien** [reply]
oh yeah, i forgot to mention, latest sdl do have libdecor support, but it's a plugin

so libdecor will have to be packaged by contributors and members of the OSU! lazer project

**Reply by RushingAlien** [reply]
Hello again! Is there somewhere I can contribute to to make changes to how the appimage is built?

**Reply by bdach** [reply]
@RushingAlien we haven't even agreed if we want to entertain some compositors' asinine insistence on having every app reimplement the titlebar which works and has worked on every other desktop platform in existence until now.

i say we continue to ignore it until they come to their senses.

**Reply by RushingAlien** [reply]
I mean, the changes involved will only be building libdecor before SDL.

And I've talked to a few GNOME devs on the matter, some are interested in implementing SSD to Mutter Wayland, but it's a big hurdle that require huge rewrites of the Mutter codebase.

And among other things, GNOME knows apps don't want to reimplement titlebar, that's why they authored libdecor as sort of a drop-in thing to start making CSD easy for apps

**Reply by bdach** [reply]
if you really wish to spend your precious time on reimplementing basic window compositor functionality because gnome is being gnome then we can probably support this if you can figure out how to shove the libdecor binaries or whatever into https://github.com/ppy/SDL2-CS/tree/master. maybe. i'm not really willing to lift a finger in that direction myself.

**Reply by RushingAlien** [reply]
https://github.com/ppy/SDL2-CS/pull/170

Hello, can this be reviewed?

**Reply by peppy** [reply]
When we find the throughput. It may not be until next year as we are currently focussing on other tasks. It's on the review list so you can rest assured it will eventually get attended to.

**Reply by RushingAlien** [reply]
Thanks :D, I can wait

**Reply by bdach** [reply]
I'll just note that proper testing will require a Linux setup with not only Wayland but also one of the compositors that ~~refuse to do their basic jobs~~ rely on client side decorations. I don't have one readily available and setting one up would take a while.

**Reply by RushingAlien** [reply]
I understand that, you can run `weston` and even `gnome-shell`(wayland) nested as an x window within an existing x.org or xwayland server. (both don't have SSD support)

To run osu! within the nested sessions, just put in the correct `WAYLAND_DISPLAY` variable to point the app to the right wayland socket.
of course, needs `SDL_VIDEODRIVER=wayland` env var first.

**Comment by peppy** [comment]
This sounds like a very weird direction. The window manager should be responsible for drawing window decorations. I don't think we would be supporting something that requires a plugin, unless it's as simple as toggling a compile time flag for SDL.

**Reply by smoogipoo** [reply]
This is just the way that Wayland works. Some compositors only support client-side decorations.

**Reply by peppy** [reply]
I understand that. I'm saying it's probably not the direction we want to go to add client-side ones on a per WM basis.

**Reply by bdach** [reply]
So I've googled around a bit and... I'm not sure what to think about gnome's decision to not provide decorations by default in any capacity. The only words I can muster right now would be along the lines of "detached from reality".

I'll leave some links if someone wants to go down the rabbit hole:
* [Gitlab thread](https://gitlab.gnome.org/GNOME/mutter/-/issues/217) where gnome devs say adding server side decorations are a wontfix
* Incidentally, that thread includes [a comment from one of the SDL maintainers](https://gitlab.gnome.org/GNOME/mutter/-/issues/217#note_355246) asking for server side support as well
* [This blog post about the "direction" of gnome](https://blogs.gnome.org/tbernard/2018/01/26/csd-initiative/) which includes the following quote, for instance:

> The process will be different for each app, and the changes required will range from “can be done in a weekend” to “holy shit we have to redesign the entire app”.

I don't know how you can develop a window manager and say that last part in earnest. Seems like delusions of grandeur to me.

**Reply by RushingAlien** [reply]
I have to mention that libdecor prioritizes SSD (WM drawn decorations) if the compositor supports it
and draws CSD (app drawn decorations) only if the compositor doesn't have the ability

**Reply by RushingAlien** [reply]
> I'm not sure what to think about gnome's decision to not provide decorations by default in any capacity. The only words I can muster right now would be along the lines of "detached from reality".

At least they aren't sitting still about it. GNOME wrote libdecor so apps can have a pluggable way to go about drawing decorations rather than writing their own.
![Cuplikanlayar_dari_2022-01-13_09-54-33](https://user-images.githubusercontent.com/80886145/154801973-7cd478cb-8044-41f4-9d2b-91849003b21f.png)
here i am running Osu with system SDL2 with libdecor installed.

also using the cairo/gtk plugin for libdecor
